File: fsharp/minesweeper/Minesweeper.fs
module Minesweeper

let addMineCounts mc1 mc2 =
    if mc1 < 0 || mc2 < 0 then -1 else mc1 + mc2

let addLists l1 l2 =
    List.zip l1 l2 
    |> List.map (fun (mc1, mc2) -> addMineCounts mc1 mc2)

let addFlowers flower1 flower2 = 
    List.zip flower1 flower2 
    |> List.map (fun (l1, l2) -> addLists l1 l2)

let singleMineFlower col cols = 
    let edge   = List.replicate col 0 @ List.replicate 3 1 @ List.replicate (cols - col - 1) 0
    let middle = List.updateAt (col + 1) -1 edge
    [edge; middle; edge]

let getFlowersForRow row cols = 
    let defaultList = List.replicate 3 (List.replicate (cols + 2) 0)
    if row |> String.exists ((=) '*') then 
        row 
        |> Seq.indexed 
        |> Seq.filter (fun (_, ch) -> ch = '*')
        |> Seq.map (fun (col, _) -> singleMineFlower col cols)
        |> Seq.reduce addFlowers
    else defaultList

let getResultString resultRow = 
    let l = List.length resultRow
    resultRow |> List.skip 1 |> List.take (l - 2)
    |> List.map (fun num -> 
            match num with 
            | -1 -> "*"
            | 0  -> " "
            | n  -> n.ToString()   
        )
    |> String.concat ""

let cleanUp result = 
    let l = List.length result
    result |> List.skip 1 |> List.take (l - 2) |> List.map getResultString

let annotate input =  
    if input = [] || input = [""] then input
    else 
        let cols = input |> List.head |> String.length 
        input 
        |> List.map (fun row -> getFlowersForRow row cols)
        |> List.reduce (fun acc rowFlowers -> 
            let accLength = acc |> List.length
            (acc.[.. accLength - 3]) @ [
                addLists acc.[accLength - 2] rowFlowers.[0];
                addLists acc.[accLength - 1] rowFlowers.[1];
                rowFlowers.[2]
            ]
            )
        |> cleanUp

File: elixir/minesweeper/lib/minesweeper.ex
defmodule Minesweeper do
  @doc """
  Annotate empty spots next to mines with the number of mines next to them.
  """

  @spec annotate([String.t()]) :: [String.t()]

  def annotate([]), do: []
  def annotate([""]), do: [""]
  def annotate(board) do
    cols = board |> Enum.at(0) |> String.length()
    board
    |> Enum.map(&get_flowers(&1, cols))
    |> Enum.reduce(fn row_flower, acc ->
      new_row_at_minus_two = add_arrs(acc |> Enum.at(-2), row_flower |> Enum.at(0))
      new_row_at_minus_one = add_arrs(acc |> Enum.at(-1), row_flower |> Enum.at(1))
      new_last_row = row_flower |> Enum.at(2)
      (acc |> Enum.drop(-2)) ++ [new_row_at_minus_two, new_row_at_minus_one, new_last_row]
    end)
    |> then(&clean_up/1)
  end

  defp get_flowers(row, cols) do
    default_list = List.duplicate(List.duplicate(0, cols + 2), 3)
    row
    |> String.codepoints
    |> Enum.with_index
    |> Enum.filter(&(elem(&1, 0) == "*"))
    |> Enum.map(fn {_, col} -> flower(col, cols) end)
    |> Enum.reduce(default_list, &add_flowers/2)
  end

  defp flower(col, cols) do
    edge = [1, 1, 1] ++ List.duplicate(0, cols - 1)
    middle = [1, -1, 1] ++ List.duplicate(0, cols - 1)
    if col == 0 do
      [edge, middle, edge]
    else
      edge = Enum.slide(edge, 0..2, col + 2)
      middle = Enum.slide(middle, 0..2, col + 2)
      [edge, middle, edge]
    end
  end

  defp add_flowers(flower1, flower2) do
    Enum.zip(flower1, flower2) |> Enum.map(fn {arr1, arr2} -> add_arrs(arr1, arr2) end)
  end

  defp add_arrs(arr1, arr2) do
    Enum.zip(arr1, arr2) |> Enum.map(fn {el1, el2} -> add_nums(el1, el2) end)
  end

  defp add_nums(el1, el2) do
    if el1 < 0 || el2 < 0 do
      -1
    else
      el1 + el2
    end
  end

  defp get_result_string(row) do
    row
    |> Enum.drop(1)
    |> Enum.drop(-1)
    |> Enum.map(fn num ->
      case num do
        -1 -> "*"
        0  -> " "
        n  -> to_string(n)
      end
    end)
    |> Enum.join
  end

  defp clean_up(result_rows) do
    result_rows |> Enum.drop(1) |> Enum.drop(-1) |> Enum.map(&get_result_string/1)
  end

end
